// israeli-bank-scrapers/lib/helpers/transactions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.filterOldTransactions = filterOldTransactions;
exports.fixInstallments = fixInstallments;
exports.sortTransactionsByDate = sortTransactionsByDate;
var _lodash = _interopRequireDefault(require("lodash"));
var _moment = _interopRequireDefault(require("moment"));
var _transactions = require("../transactions");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function isNormalTransaction(txn) {
  return txn && txn.type === _transactions.TransactionTypes.Normal;
}
function isInstallmentTransaction(txn) {
  return txn && txn.type === _transactions.TransactionTypes.Installments;
}
function isNonInitialInstallmentTransaction(txn) {
  return isInstallmentTransaction(txn) && !!txn.installments && txn.installments.number > 1;
}
function isInitialInstallmentTransaction(txn) {
  return isInstallmentTransaction(txn) && !!txn.installments && txn.installments.number === 1;
}
function fixInstallments(txns) {
  return txns.map(txn => {
    const clonedTxn = _objectSpread({}, txn);
    if (isInstallmentTransaction(clonedTxn) && isNonInitialInstallmentTransaction(clonedTxn) && clonedTxn.installments) {
      const dateMoment = (0, _moment.default)(clonedTxn.date);
      const actualDateMoment = dateMoment.add(clonedTxn.installments.number - 1, 'month');
      clonedTxn.date = actualDateMoment.toISOString();
    }
    return clonedTxn;
  });
}
function sortTransactionsByDate(txns) {
  return _lodash.default.sortBy(txns, ['date']);
}
function filterOldTransactions(txns, startMoment, combineInstallments) {
  return txns.filter(txn => {
    const combineNeededAndInitialOrNormal = combineInstallments && (isNormalTransaction(txn) || isInitialInstallmentTransaction(txn));
    return !combineInstallments && startMoment.isSameOrBefore(txn.date) || combineNeededAndInitialOrNormal && startMoment.isSameOrBefore(txn.date);
  });
}
