// israeli-bank-scrapers/lib/helpers/waiting.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TimeoutError = exports.SECOND = void 0;
exports.raceTimeout = raceTimeout;
exports.runSerial = runSerial;
exports.sleep = sleep;
exports.waitUntil = waitUntil;
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.promise.js");
class TimeoutError extends Error {}
exports.TimeoutError = TimeoutError;
const SECOND = exports.SECOND = 1000;
function timeoutPromise(ms, promise, description) {
  const timeout = new Promise((_, reject) => {
    const id = setTimeout(() => {
      clearTimeout(id);
      const error = new TimeoutError(description);
      reject(error);
    }, ms);
  });
  return Promise.race([promise,
  // casting to avoid type error- safe since this promise will always reject
  timeout]);
}

/**
 * Wait until a promise resolves with a truthy value or reject after a timeout
 */
function waitUntil(asyncTest, description = '', timeout = 10000, interval = 100) {
  const promise = new Promise((resolve, reject) => {
    function wait() {
      asyncTest().then(value => {
        if (value) {
          resolve(value);
        } else {
          setTimeout(wait, interval);
        }
      }).catch(() => {
        reject();
      });
    }
    wait();
  });
  return timeoutPromise(timeout, promise, description);
}
function raceTimeout(ms, promise) {
  return timeoutPromise(ms, promise, 'timeout').catch(err => {
    if (!(err instanceof TimeoutError)) throw err;
  });
}
function runSerial(actions) {
  return actions.reduce((m, a) => m.then(async x => [...x, await a()]), Promise.resolve(new Array()));
}
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
